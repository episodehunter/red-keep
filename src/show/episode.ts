export const EpisodeDefinition = `
  type Episode {
    id: ID!,
    name: String!
  }
`

export const EpisodeResolver = {
  episodes: (obj: { id: number }, args: {}, context: any) => {
    console.log('episodes obj', obj)
    console.log('episodes args', args)
    console.log('episodes context', context)
    return [{ id: 5, name: 'First episode' }]
  }
}
