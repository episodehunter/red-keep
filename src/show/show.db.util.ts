import { Maybe } from 'monet'
import { BadInput, ShowExistError, ShowNotExistError } from '../custom-error'
import { Db } from '../types/context.type'
import { addNewEpisodesInDb } from '../episode/episode.db.util'
import { ShowDefinitionType } from '../root-type'
import {
  getShowId,
  mapDefinitionToDatabaseShow,
  mapDatabaseShowToDefinition,
  mapDatabaseShowIdsToDefinition
} from './show.resolve.util'

const showTableName = 'tv_show'

function findShowId(db: Db, show: Partial<ShowDefinitionType>): Promise<number | null> {
  return db
    .first('id')
    .from(showTableName)
    .where(getShowId(show))
    .then(row => {
      if (row) {
        return row.id || null
      }
      return null
    }) as any
}

export function assertShowExist(
  db: Db,
  show: Partial<ShowDefinitionType>
): Promise<number> {
  return findShowId(db, show).then(id => {
    if (!id) {
      throw new ShowNotExistError('Show do not exist')
    }
    return id
  })
}

export function assertShowNotExist(
  db: Db,
  show: Partial<ShowDefinitionType>
): Promise<null> {
  return findShowId(db, show).then(id => {
    if (id) {
      throw new ShowExistError('Show already exist')
    }
    return null
  })
}

export function updateShow(
  db: Db,
  id: number,
  show: Partial<ShowDefinitionType>
): Promise<number> {
  return Promise.resolve(mapDefinitionToDatabaseShow(show))
    .then(dbShow =>
      db(showTableName)
        .update(dbShow)
        .where('id', id)
    )
    .then(() => id)
}

export function addShow(db: Db, show: ShowDefinitionType): Promise<number[]> {
  if (show.id) {
    throw new BadInput('Can not set id for insert!')
  } else if (!show.tvdbId) {
    throw new BadInput('tvdbId is required!')
  }
  return assertShowNotExist(db, show)
    .then(() => mapDefinitionToDatabaseShow(show))
    .then(dbShow =>
      db.transaction(trx =>
        trx(showTableName)
          .insert(dbShow)
          .then(insertIds => insertIds[0])
          .then(showId => addNewEpisodesInDb(trx, showId, show.episodes))
      )
    )
}

export async function findShow(
  db: Db,
  showIds: { id?: number; tvdbId?: number; imdbId?: string }
): Promise<ShowDefinitionType | null> {
  return Maybe.fromNull(
    await db
      .first('*')
      .from(showTableName)
      .where(getShowId(showIds))
  )
    .map(mapDatabaseShowToDefinition)
    .orNull()
}

export function filterOutNonExistingShows(db: Db, tvdbIds: number[]) {
  return Promise.resolve(tvdbIds.map(id => id | 0).filter(id => id > 0))
    .then(ids =>
      db
        .select('id', 'tvdb_id', 'imdb_id')
        .from(showTableName)
        .whereIn('tvdb_id', ids)
    )
    .then(ids => ids.map(mapDatabaseShowIdsToDefinition))
}
