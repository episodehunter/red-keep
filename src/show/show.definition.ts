export const ShowDefinition = `
  enum WeekDay {
    Monday
    Tuesday
    Wednesday
    Thursday
    Friday
    Saturday
    Sunday
  }

  type ShowIds {
    id: Int!,
    tvdbId: Int!,
    imdbId: String
  }

  type Show {
    id: Int!,
    tvdbId: Int!,
    imdbId: String,
    name: String!,
    airsDayOfWeek: WeekDay,
    airsTime: String,
    firstAired: String,
    genre: [String],
    language: String,
    network: String,
    overview: String,
    runtime: Int,
    ended: Boolean,
    fanart: String,
    poster: String,
    lastupdate: Int,
    episodes(onlyMissingImages: Boolean): [Episode]
  }

  input ShowInput {
    id: Int,
    tvdbId: Int,
    imdbId: String,
    name: String!,
    airsDayOfWeek: WeekDay,
    airsTime: String,
    firstAired: String,
    genre: [String],
    language: String,
    network: String,
    overview: String,
    runtime: Int,
    ended: Boolean!,
    fanart: String,
    poster: String,
    lastupdate: Int!,
    episodes: [EpisodeInput]
  }
`
