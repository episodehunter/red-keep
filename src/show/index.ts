import { ShowDefinition } from './show.definition'
import { EpisodeDefinition } from './episode/episode.definition'
export { ShowResolver } from './show.resolver'

export const ShowDefinitions = [ShowDefinition, EpisodeDefinition]
