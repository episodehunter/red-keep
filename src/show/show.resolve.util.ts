import { ShowStatusType, ShowDatabaseType, ShowDefinitionType } from './show.type'
import { safeStringSplit } from '../util'

export function convertStatusBoleanToString(ended?: boolean): ShowStatusType | undefined {
  if (typeof ended === 'boolean') {
    return ended ? 'Ended' : 'Continuing'
  }
  return undefined
}

export function convertGenreArrayToString(genre: string[]) {
  return Array.isArray(genre) ? genre.join('|') : undefined
}

export function getShowId(
  ids: Partial<{
    id: number
    tvdb_id: number
    imdb_id: string
    tvdbId: number
    imdbId: string
  }>
) {
  const { id, tvdb_id, imdb_id, imdbId, tvdbId } = ids
  if (id) {
    return { id }
  } else if (tvdb_id || tvdbId) {
    return { tvdb_id: tvdb_id || tvdbId }
  } else if (imdb_id || imdbId) {
    return { imdb_id: imdb_id || imdbId }
  }
  throw new Error('Missing id for updating show')
}

export function mapDatabaseShowToDefinition(show: ShowDatabaseType): ShowDefinitionType {
  return {
    id: show.id,
    tvdbId: show.tvdb_id,
    imdbId: show.imdb_id,
    name: show.name,
    airsDayOfWeek: show.airs_dayOfWeek,
    airsTime: show.airs_time,
    firstAired: show.first_aired,
    genre: safeStringSplit(show.genre, '|'),
    language: show.language,
    network: show.network,
    overview: show.overview,
    runtime: show.runtime,
    ended: show.status === 'Ended',
    fanart: show.fanart,
    poster: show.poster,
    lastupdate: show.lastupdate
  }
}

export function mapDefinitionToDatabaseShow(
  show: Partial<ShowDefinitionType>
): Partial<ShowDatabaseType> {
  return {
    id: show.id,
    tvdb_id: show.tvdbId,
    imdb_id: show.imdbId,
    name: show.name,
    airs_dayOfWeek: show.airsDayOfWeek,
    airs_time: show.airsTime,
    first_aired: show.firstAired,
    genre: convertGenreArrayToString(show.genre),
    language: show.language,
    network: show.network,
    overview: show.overview,
    runtime: show.runtime,
    status: convertStatusBoleanToString(show.ended),
    fanart: show.fanart,
    poster: show.poster,
    lastupdate: show.lastupdate
  }
}

export function assertValue(msg: string) {
  return (value?: any) => {
    if (value) {
      return value
    }
    throw new Error(msg)
  }
}
