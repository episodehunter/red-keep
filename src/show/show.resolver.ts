import { Maybe } from 'monet'
import { Context } from '../types/context.type'
import { EpisodeResolver } from './episode/episode.resolver'
import { ShowDatabaseType, ShowDefinitionType } from './show.type'
import { safeStringSplit } from '../util'

function mapDatabaseShowToDefinition(show: ShowDatabaseType): ShowDefinitionType {
  return {
    id: show.id,
    tvdbId: show.tvdb_id,
    imdbId: show.imdb_id,
    name: show.name,
    airsDayOfWeek: show.airs_dayOfWeek,
    airsTime: show.airs_time,
    firstAired: show.first_aired,
    genre: safeStringSplit(show.genre, '|'),
    language: show.language,
    network: show.network,
    overview: show.overview,
    runtime: Number(show.runtime),
    ended: show.status === 'Ended',
    fanart: show.fanart,
    poster: show.poster,
    lastupdate: show.lastupdate
  }
}

export const ShowResolver = {
  RootQuery: {
    show: async (
      obj: void,
      args: { id: number },
      context: Context
    ): Promise<ShowDefinitionType | null> => {
      return Maybe.fromNull(
        await context.db
          .first('*')
          .from('tv_show')
          .where('id', args.id)
      )
        .map(mapDatabaseShowToDefinition)
        .orNull()
    }
  },
  Show: {
    episodes: EpisodeResolver.episodes
  }
}
