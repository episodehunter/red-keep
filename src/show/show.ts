import { Maybe } from 'monet'
import { Context } from '../types/context.type'
import { EpisodeDefinition, EpisodeResolver } from './episode'

export const ShowDefinition = `
  enum WeekDay {
    Monday
    Tuesday
    Wednesday
    Thursday
    Friday
    Saturday
    Sunday
  }

  enum ShowStatus {
    Ended
    Continuing
  }

  type Show {
    id: ID!,
    tvdbId: Int!,
    imdbId: String,
    name: String!,
    airsDayOfWeek: WeekDay,
    airsTime: String,
    firstAired: String,
    genre: [String],
    language: String,
    network: String,
    overview: String,
    runtime: Int,
    ended: Boolean,
    fanart: String,
    poster: String,
    lastupdate: Int,
    episodes: [Episode]
  }
`

type ShowStatus = 'Continuing' | 'Ended'

type ShowDatabase = {
  id: number
  tvdb_id: number
  imdb_id: string
  name: string
  airs_dayOfWeek: string
  airs_time: string
  first_aired: string
  genre: string
  language: string
  network: string
  overview: string
  runtime: string
  status: ShowStatus
  fanart: string
  poster: string
  lastupdate: number
}

type ShowDefinition = {
  id: number
  tvdbId: number
  imdbId: string
  name: string
  airsDayOfWeek: string
  airsTime: string
  firstAired: string
  genre: string[]
  language: string
  network: string
  overview: string
  runtime: number
  ended: boolean
  fanart: string
  poster: string
  lastupdate: number
}

function safeStringSplit(str: string, key: string) {
  if (!str || !str.split) {
    return []
  }
  return str.split(key)
}

function mapDatabaseShowToDefinition(show: ShowDatabase): ShowDefinition {
  return {
    id: show.id,
    tvdbId: show.tvdb_id,
    imdbId: show.imdb_id,
    name: show.name,
    airsDayOfWeek: show.airs_dayOfWeek,
    airsTime: show.airs_time,
    firstAired: show.first_aired,
    genre: safeStringSplit(show.genre, '|'),
    language: show.language,
    network: show.network,
    overview: show.overview,
    runtime: Number(show.runtime),
    ended: show.status === 'Ended',
    fanart: show.fanart,
    poster: show.poster,
    lastupdate: show.lastupdate
  }
}

export const ShowResolver = {
  RootQuery: {
    show: async (
      obj: any,
      args: { id: number },
      context: Context
    ): Promise<ShowDefinition | null> => {
      return Maybe.fromNull(
        await context.db
          .first('*')
          .from('tv_show')
          .where('id', args.id)
          .then(result => result || null)
      )
        .map(mapDatabaseShowToDefinition)
        .orNull()
    }
  },
  Show: {
    episodes: EpisodeResolver.episodes
  }
}

export const ShowDefinitions = [ShowDefinition, EpisodeDefinition]
