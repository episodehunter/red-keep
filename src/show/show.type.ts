import { EpisodeDefinitionType } from './episode/episode.type'
export type ShowStatusType = 'Continuing' | 'Ended'

export type ShowDatabaseType = {
  id: number
  tvdb_id: number
  imdb_id: string
  name: string
  airs_dayOfWeek: string
  airs_time: string
  first_aired: string
  genre: string
  language: string
  network: string
  overview: string
  runtime: number
  status: ShowStatusType
  fanart: string
  poster: string
  lastupdate: number
}

export type ShowDefinitionType = {
  id: number
  tvdbId: number
  imdbId: string
  name: string
  airsDayOfWeek: string
  airsTime: string
  firstAired: string
  genre: string[]
  language: string
  network: string
  overview: string
  runtime: number
  ended: boolean
  fanart: string
  poster: string
  episodes?: EpisodeDefinitionType[]
  lastupdate: number
}
