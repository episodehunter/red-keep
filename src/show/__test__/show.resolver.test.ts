import { create } from 'chain-spy'
import { ShowResolver } from '../show.resolver'

test('Resolve a show', async () => {
  // Arrange
  const dbShow = {
    id: 5,
    tvdb_id: 2,
    imdb_id: 'tt12345',
    name: 'Dexter',
    airs_dayOfWeek: 'Monday',
    airs_time: '09:00 PM',
    first_aired: '2016-03-15',
    genre: 'Drama|Action',
    language: 'en',
    network: 'AMC',
    overview: 'Something, somthing',
    runtime: '60',
    ended: 'Ended',
    fanart: 'fanart.jpg',
    poster: 'poster.jpg',
    lastupdate: 1234567891
  }
  const dbProxy = create({ where: () => dbShow })
  const args = { id: 5 }

  // Act
  const result = await ShowResolver.RootQuery.show(undefined, args, { db: dbProxy })

  // Assert
  expect(result).toMatchSnapshot()
  expect(dbProxy.__execution_log__).toMatchSnapshot()
})

test('Resolve null', async () => {
  // Arrange
  const dbShow = null
  const dbProxy = create({ where: () => dbShow })
  const args = { id: 5 }

  // Act
  const result = await ShowResolver.RootQuery.show(undefined, args, { db: dbProxy })

  // Assert
  expect(result).toBe(null)
})
