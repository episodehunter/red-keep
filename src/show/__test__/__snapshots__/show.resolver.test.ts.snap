// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resolve a show 1`] = `
Object {
  "airsDayOfWeek": "Monday",
  "airsTime": "09:00 PM",
  "ended": false,
  "fanart": "fanart.jpg",
  "firstAired": "2016-03-15",
  "genre": Array [
    "Drama",
    "Action",
  ],
  "id": 5,
  "imdbId": "tt12345",
  "language": "en",
  "lastupdate": 1234567891,
  "name": "Dexter",
  "network": "AMC",
  "overview": "Something, somthing",
  "poster": "poster.jpg",
  "runtime": 60,
  "tvdbId": 2,
}
`;

exports[`Resolve a show 2`] = `
Array [
  Object {
    "name": "first",
    "type": "get",
  },
  Object {
    "args": Array [
      "*",
    ],
    "type": "apply",
  },
  Object {
    "name": "from",
    "type": "get",
  },
  Object {
    "args": Array [
      "tv_show",
    ],
    "type": "apply",
  },
  Object {
    "name": "where",
    "type": "get",
  },
  Object {
    "args": Array [
      "id",
      5,
    ],
    "type": "apply",
  },
]
`;
